webapp/src/blocks.tsx

```diff
@@ -789,68 +802,472 @@ export class Editor extends srceditor.Editor {
+            if (subns) return [];
+            let extraBlocks: (toolbox.BlockDefinition | toolbox.ButtonDefinition)[] = [];
+            const onStartNamespace = pxt.appTarget.runtime.onStartNamespace || "loops";
+            if (ns == onStartNamespace) {
+                extraBlocks.push({
+                    name: ts.pxtc.ON_START_TYPE,
+                    attributes: {
+                        blockId: ts.pxtc.ON_START_TYPE,
+                        weight: pxt.appTarget.runtime.onStartWeight || 10
+                    },
+                    blockXml: `<block type="pxt-on-start"></block>`,
+                    noNamespace: true
+                });
+            }
+            // Inject pause until block
+            const pauseUntil = snippets.getPauseUntil();
+            if (pauseUntil && ns == pauseUntil.attributes.blockNamespace) {
+                extraBlocks.push(pauseUntil);
+            }
+            // Add extension buttons
+            if (!subns) {
+                that.extensions.forEach(config => {
+                    const name = config.name;
+                    const namespace = config.extension.namespace || name;
+                    if (ns == namespace) {
+                        extraBlocks.push({
+                            name: `EXT${name}_BUTTON`,
+                            type: "button",
+                            attributes: {
+                                blockId: `EXT${name}_BUTTON`,
+                                label: config.extension.label ? Util.rlf(config.extension.label) : Util.lf("Editor"),
+                                weight: 101
+                            },
+                            callback: () => {
+                                that.openExtension(name);
+                            }
+                        });
+                    }
+                })
+            }
+            return extraBlocks;
+        }
+
+        if (!snippets.isBuiltin(ns)) {
+            return getNonBuiltInBlocks(ns).concat(getExtraBlocks(ns, subns));
+        } else {
+            return getBuiltInBlocks(ns).concat(getExtraBlocks(ns, subns));
+        }
+    }
+
+    private flyoutBlockXmlCache: pxt.Map<Element[]> = {};
+    public showFlyout(treeRow: toolbox.ToolboxCategory) {
+        const { nameid, subns, icon, color, name, labelLineWidth } = treeRow;
+        const ns = nameid;
+
+        if (ns == 'search') {
+            this.showSearchFlyout();
+            return;
+        }
+
+        let xmlList: Element[] = [];
+        if (this.flyoutBlockXmlCache[ns + subns]) {
+            pxt.debug("showing flyout with blocks from flyout blocks xml cache");
+            xmlList = this.flyoutBlockXmlCache[ns + subns];
+            this.showFlyoutInternal_(xmlList);
+            return;
+        }
+
+        function createHeadingLabel() {
+            const categoryName = name ? name :
+                `${subns ? `${Util.capitalize(ns)} > ${Util.capitalize(subns)}` : Util.capitalize(ns)}`;
```

Comment:
The outer template string seems unnecessary

Reply:
True